AGENTS.md

```diff
@@ -0,0 +1,247 @@
+- **Configuration**: Pydantic-based `Config` class with environment variable support
+- **Transport**: HTTPX-based with proxy support
+- **Authentication**: API tokens or JWT with automatic refresh
+
+### Key Modules Structure
+
+```text
+infrahub_sdk/
+├── client.py           # Main client implementations
+├── config.py          # Configuration management with Pydantic
+├── node/              # Node system (core data model)
+│   ├── node.py        # InfrahubNode and InfrahubNodeSync
+│   ├── attribute.py   # Node attributes
+│   └── relationship.py # Relationship management
+├── ctl/               # CLI commands (infrahubctl)
+├── pytest_plugin/    # Custom pytest plugin for Infrahub testing
+└── protocols.py       # Generated protocol classes
+```
+
+### Node System Design
+
+- **Lazy Loading**: Nodes load attributes and relationships on demand
+- **Batch Operations**: Support for bulk create/update/delete operations
+- **Relationship Management**: Automatic handling of node relationships with add/remove/replace operations
+- **Validation**: Built-in data validation with GraphQL query generation
+
+## Code Conventions & Standards
+
+### Type Hints
+
+- Use type hints for all function signatures
+- Prefer `from __future__ import annotations` for forward references
+- Use `TYPE_CHECKING` for imports only needed for type checking
+- Always run `uv run invoke format lint` after modifying some python code
+- No need to clean up the dependencies, `invoke format` will do it for you
+
+### Async/Sync Pattern
+
+All operations follow dual implementation pattern:
+
+```python
+# Async version (default)
+client = InfrahubClient()
+node = await client.get(kind="NetworkDevice")
+await node.save()
+
+# Sync version
+client = InfrahubClientSync()
+node = client.get(kind="NetworkDevice")
+node.save()
+```
+
+### Code Quality Tools
+
+- **Ruff**: Comprehensive linting and formatting (0.11.0)
+  - Configuration in `pyproject.toml` under `[tool.ruff]`
+  - Line length: 120 (formatter), 150 (pycodestyle)
+  - Max complexity: 17
+- **mypy**: Type checking with strict configuration
+  - `disallow_untyped_defs = true`
+  - Module-specific overrides in `pyproject.toml`
+- **yamllint**: YAML file validation
+- **markdownlint**: Documentation consistency
+- **Vale**: Documentation style checking
+
+### Configuration Management
+
+- Environment variables prefixed with `INFRAHUB_`
+- Proxy configuration: `INFRAHUB_PROXY` (single) or `INFRAHUB_PROXY_MOUNTS_HTTP`/`INFRAHUB_PROXY_MOUNTS_HTTPS` (separate)
+- Mutual exclusivity validation between proxy configuration methods
+
+## Infrahub-Specific Patterns
+
+### Checks Implementation
+
+**CRITICAL**: Use `validate()` method, NOT `check()`
+
+```python
+# CORRECT
+class MyCheck(InfrahubCheck):
+    def validate(self, data):  # Must be validate(), not check()
+        # validation logic
+        pass
+
+# WRONG
+class MyCheck(InfrahubCheck):
+    def check(self, data):  # This will NOT work
+        pass
```

Comment:
In all of the examples provided to the agents I think we should use methods with full typing on all parameters and return values.